src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/CollectionModelBinder.cs

```diff
@@ -164,12 +161,20 @@ protected object CreateInstance(Type targetType)
                 };
 
                 object boundValue = null;
-                var result =
-                    await bindingContext.OperationBindingContext.ModelBinder.BindModelAsync(innerBindingContext);
-                if (result != null && result.IsModelSet)
+
+                using (bindingContext.PushContext(
+                    elementMetadata,
+                    fieldName: bindingContext.FieldName,
+                    modelName: bindingContext.ModelName,
+                    model: null))
```

Comment:
Well, if we need such strong guarantees, something is wrong elsewhere. But you're right it shouldn't be as big a deal inside the loop now.